ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -661,30 +679,40 @@ IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, long deletionTimeMs) th
    * @return the {@link IndexValue} of the delete record
    * @throws StoreException
    */
-  private IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, MessageInfo info, long deletionTimeMs)
-      throws StoreException {
+   IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, MessageInfo info, long deletionTimeMs,
+      short lifeVersion) throws StoreException {
+    boolean hasLifeVersion = lifeVersion > (short) -1;
     validateFileSpan(fileSpan, true);
+    //TODO IndexValue findLatest, can be either PUT, DELETE, or UNDELETE
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
-      throw new StoreException("Id " + id + " already deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+    } else if (value != null) {
+      if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        throw new StoreException("Id " + id + " already deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+      } else if (hasLifeVersion && value.getLifeVersion() > lifeVersion) {
+        throw new StoreException("LifeVersion conflict in index. Id " + id + " LifeVersion: " + value.getLifeVersion()
+            + " Delete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
+      }
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     IndexValue newValue;
     if (value == null) {
       // It is possible that the PUT has been cleaned by compaction
+      lifeVersion = lifeVersion < 0 ? 0 : lifeVersion;
       newValue =
-          new IndexValue(size, fileSpan.getStartOffset(), info.getExpirationTimeInMs(), info.getOperationTimeMs(),
-              info.getAccountId(), info.getContainerId());
+          new IndexValue(size, fileSpan.getStartOffset(), IndexValue.FLAGS_DEFAULT_VALUE, info.getExpirationTimeInMs(),
+              info.getOperationTimeMs(), info.getAccountId(), info.getContainerId(), lifeVersion);
       newValue.clearOriginalMessageOffset();
     } else {
+      lifeVersion = hasLifeVersion ? lifeVersion : value.getLifeVersion();
```

Comment:
Use info.getLifeVersion() would be safe. Reasons stated below.

Reply:
see above comment